tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,557 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "mlir-hlo/utils/cycle_detector.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "tensorflow/compiler/mlir/disc/disc_util.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+#include "tensorflow/compiler/mlir/disc/transforms/shape_utils.h"
+
+#define DEBUG_TYPE "disc-dot-merge"
+
+namespace mlir {
+namespace disc_ral {
+namespace {
+
+llvm::Optional<int32_t> TryMergeNode(GraphCycles* graph_cycles, int32_t a,
+                                     int32_t b) {
+  if (graph_cycles == nullptr) {
+    return llvm::None;
+  }
+  bool has_edge_inserted_a2b = false;
+  if (!graph_cycles->HasEdge(a, b) && !graph_cycles->HasEdge(b, a)) {
+    has_edge_inserted_a2b = graph_cycles->InsertEdge(a, b);
+    if (!has_edge_inserted_a2b) {
+      // Cannot merge a and b as we cannot even insert an edge between a and b.
+      return llvm::None;
+    }
+  }
+  int32_t from = graph_cycles->HasEdge(a, b) ? a : b;
+  int32_t to = (from == a) ? b : a;
+  auto result = graph_cycles->ContractEdge(from, to);
+  if (!result.hasValue() && has_edge_inserted_a2b) {
+    // Restore the graph.
+    graph_cycles->RemoveEdge(a, b);
+  }
+  return result;
+}
+
+// Arrange the insert-point of `op`'s operands in the same block. Do not deal
+// with cross-block operands.
+void arrangeOperandsInsertPointInBlock(Operation* op) {
+  for (auto operand : op->getOperands()) {
+    auto operandOp = operand.getDefiningOp();
+    if ((operandOp != nullptr) && !operandOp->isBeforeInBlock(op)) {
+      operandOp->moveBefore(op);
+      arrangeOperandsInsertPointInBlock(operandOp);
+    }
+  }
+}
+
+class DotBatchingConverter {
+ public:
+  DotBatchingConverter(FuncOp func) : func_(func){};
+  void run();
+
+ private:
+  struct DotShape {
+    DimValue m_dim;
+    DimValue n_dim;
+    DimValue contracting_dim;
+    SmallVector<DimValue> batching_dims;
+    mhlo::DotDimensionNumbersAttr dimension_numbers;
+    Type element_type;
+    bool operator==(const DotShape& other) const {
+      return (m_dim == other.m_dim) && (n_dim == other.n_dim) &&
+             (contracting_dim == other.contracting_dim) &&
+             (batching_dims == other.batching_dims) &&
+             (dimension_numbers == other.dimension_numbers) &&
+             (element_type == other.element_type);
+    }
+  };
```

Comment:
Is this actually the same as requiring: 
'''
(lhs().getType() == other.lhs().getType() &&
rhs().getType() == other.rhs().getType() &&
dimension_numbers() == other.dimension_numbers())
'''